src/storage/s3.rs

```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+        let res = self
+            .client
+            .get_object(GetObjectRequest {
+                bucket: self.bucket.to_string(),
+                key: path.into(),
+                ..Default::default()
+            })
+            .sync()?;
+
+        let mut b = res.body.unwrap().into_blocking_read();
+        let mut content = Vec::with_capacity(
+            res.content_length
+                .and_then(|l| l.try_into().ok())
+                .unwrap_or(0),
+        );
+        b.read_to_end(&mut content).unwrap();
+
+        let date_updated = parse_timespec(&res.last_modified.unwrap())?;
+
+        Ok(Blob {
+            path: path.into(),
+            mime: res.content_type.unwrap(),
+            date_updated,
+            content,
+        })
+    }
+
+    pub(super) fn store_batch(&mut self, batch: &[Blob]) -> Result<(), Error> {
+        use futures::stream::FuturesUnordered;
+        use futures::stream::Stream;
+        let mut attempts = 0;
+
+        loop {
+            let mut futures = FuturesUnordered::new();
+            for blob in batch {
+                futures.push(
+                    self.client
+                        .put_object(PutObjectRequest {
+                            bucket: self.bucket.to_string(),
+                            key: blob.path.clone(),
+                            body: Some(blob.content.clone().into()),
+                            content_type: Some(blob.mime.clone()),
+                            ..Default::default()
+                        })
+                        .inspect(|_| {
+                            crate::web::metrics::UPLOADED_FILES_TOTAL.inc_by(1);
+                        }),
+                );
+            }
+            attempts += 1;
+
+            match self.runtime.block_on(futures.map(drop).collect()) {
+                // this batch was successful, start another batch if there are still more files
+                Ok(_) => break,
+                Err(err) => {
+                    error!("failed to upload to s3: {:?}", err);
+                    // if a futures error occurs, retry the batch
+                    if attempts > 2 {
+                        panic!("failed to upload 3 times, exiting");
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+// public for testing
+pub(crate) const TIME_FMT: &str = "%a, %d %b %Y %H:%M:%S %Z";
+
+fn parse_timespec(raw: &str) -> Result<Timespec, Error> {
+    Ok(time::strptime(raw, TIME_FMT)?.to_timespec())
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use crate::test::*;
+
+    use crate::storage::s3::S3Backend;
+    use rusoto_s3::{
+        CreateBucketRequest, DeleteBucketRequest, DeleteObjectRequest, ListObjectsRequest, S3,
+    };
+
+    use std::cell::RefCell;
+    use std::slice;
+
+    pub(crate) struct TestS3(RefCell<S3Backend<'static>>);
```

Comment:
I'd move this to `src/tests/s3.rs`.

Reply:
Moving this to the tests module requires making almost everything in `src/storage/s3.rs` `pub(crate)`. I'm ok doing that if you want me to but I think the current layout gives a better idea of what functions are called where.

Reply:
Can we do a `src/storage/s3_test.rs` instead?

Reply:
I moved it to `src/storage/s3/test_s3.rs` so it had access to all of `s3.rs`, does that work?

Reply:
I ended up making it `test.rs` and renaming the inner module to `tests` to avoid a naming conflict.